constraint-solver/src/solver/linearized.rs

```diff
@@ -0,0 +1,449 @@
+    fn retain_variables(&mut self, variables_to_keep: &HashSet<Variable<V>>) {
+        // There are constraints that only contain `Variable::Linearized` that
+        // connect quadratic terms with the original constraints. We could try to find
+        // those, but let's just keep all of them for now.
+        let mut variables_to_keep = variables_to_keep.clone();
+        variables_to_keep.extend(self.var_dispenser.all_dispensed_vars());
+        self.solver.retain_variables(&variables_to_keep);
+    }
+
+    fn range_constraint_for_expression(
+        &self,
+        expr: &GroupedExpression<T, Variable<V>>,
+    ) -> RangeConstraint<T::FieldType> {
+        // Ask the solver directly for the range constraint of the expression.
+        let direct = self.solver.range_constraint_for_expression(expr);
+        // See if we have a direct substitution for the expression by a variable.
+        let simple_substituted = self
+            .linearizer
+            .try_substitute_by_existing_var(expr)
+            .map(|expr| self.solver.range_constraint_for_expression(&expr))
+            .unwrap_or_default();
+        // Try to re-do the linearization
+        let substituted = self
+            .linearizer
+            .try_linearize_existing(expr.clone())
+            .map(|expr| self.solver.range_constraint_for_expression(&expr))
+            .unwrap_or_default();
+        direct
+            .conjunction(&simple_substituted)
+            .conjunction(&substituted)
+    }
+
+    fn are_expressions_known_to_be_different(
+        &mut self,
+        a: &GroupedExpression<T, Variable<V>>,
+        b: &GroupedExpression<T, Variable<V>>,
+    ) -> bool {
+        let a = iter::once(a.clone()).chain(
+            (!a.is_affine())
+                .then(|| self.linearizer.try_linearize_existing(a.clone()))
+                .flatten(),
+        );
+        let b = iter::once(b.clone()).chain(
+            (!b.is_affine())
+                .then(|| self.linearizer.try_linearize_existing(b.clone()))
+                .flatten(),
+        );
+        a.cartesian_product(b)
+            .any(|(a, b)| self.solver.are_expressions_known_to_be_different(&a, &b))
+    }
+}
+
+struct Linearizer<T, V> {
+    constraints_to_add: Vec<GroupedExpression<T, V>>,
+    substitutions: HashMap<GroupedExpression<T, V>, V>,
+}
+
+impl<T, V> Default for Linearizer<T, V> {
+    fn default() -> Self {
+        Linearizer {
+            constraints_to_add: vec![],
+            substitutions: HashMap::new(),
+        }
+    }
+}
+
+impl<T: RuntimeConstant + Hash, V: Clone + Eq + Ord + Hash> Linearizer<T, V> {
+    /// Linearizes the constraint by introducing new variables for
+    /// non-affine parts. The new constraints are appended to
+    /// `self.constraints_to_add` and must be added to the system.
+    /// The linearized expression is returned.
+    fn linearize(
+        &mut self,
+        expr: GroupedExpression<T, V>,
+        var_dispenser: &mut impl FnMut() -> V,
```

Comment:
```suggestion
        var_dispenser: &mut LinearizedVarDispenser,
```
Needs less explanation and simplifies usage.

Reply:
Or, maybe better, shouldn't the Linearizer own the var dispenser?